pkg/notifier/webhook.go

```diff
@@ -0,0 +1,122 @@
+package notifier
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/sha256"
+	"crypto/tls"
+	"encoding/hex"
+	"encoding/json"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+type webhookNotifier struct {
+	WebhookURL         string
+	WebhookURLAllClear string
+	WebhookSecret      string
+	AllowInsecureTLS   bool
+}
+
+func ComputeHmacSha256(secret string, payload []byte) string {
+	key := []byte(secret)
+	h := hmac.New(sha256.New, key)
+	h.Write([]byte(payload))
+	return hex.EncodeToString(h.Sum(nil))
+}
+
+// NewWebhook creates a new webhook notifier from the supplied configuration.
+func NewWebhook(WebhookURL string,
+	WebhookURLAllClear string,
+	WebhookSecret string,
+	AllowInsecureTLS bool) (Notifier, error) {
+
+	if WebhookURL == "" {
+		return nil, errors.New("Unable to initialize webhook: webhookURL is empty")
+	}
+	if WebhookURLAllClear == "" {
+		return nil, errors.New("Unable to initialize webhook: webhookURL_all_clear is empty")
+	}
+
+	return &webhookNotifier{
+		WebhookURL,
+		WebhookURLAllClear,
+		WebhookSecret,
+		AllowInsecureTLS,
+	}, nil
+}
+
+// Notify implements the Notifier interface for webhook.
+func (w *webhookNotifier) Notify(msg Message) error {
+	postBody, _ := json.Marshal(map[string]interface{}{
+		"message": msg.Format(),
+		"meta":    msg.Meta,
+	})
+	request, err := http.NewRequest("POST", w.WebhookURL, bytes.NewBuffer(postBody))
+	request.Header.Set("Content-Type", "application/json")
+	request.Header.Set("X-Program", msg.Program)
+
+	if w.WebhookSecret != "" {
+		timestamp := strconv.FormatInt(time.Now().Unix(), 10)
+		payload := append([]byte(timestamp), postBody...)
+		signature := ComputeHmacSha256(w.WebhookSecret, payload)
+
+		request.Header.Set("X-Timestamp", timestamp)
+		request.Header.Set("X-HMAC-SHA256", signature)
+	}
+
+	client := &http.Client{Timeout: time.Second * 10}
```

Comment:
The http client should be created once and re-used, there are 2 options.
Pass created http client from root.go into `func NewWebhook`, or create it in `func NewWebhook` based on config.

This is required to allow the http client to re-use TCP connections and connection pooling. https://golang.org/pkg/net/http/#Client

Reply:
The http.Client is now created in `func NewWebhook` based on the config.